File: rust/crates/symphony-tracker/src/tracker.rs
use async_trait::async_trait;
use symphony_core::Issue;

use crate::errors::TrackerError;

/// Minimal issue-state record returned by reconciliation refresh
/// (`fetch_issue_states_by_ids`).
#[derive(Debug, Clone)]
pub struct IssueState {
    pub id: String,
    pub identifier: String,
    pub state: String,
}

/// SPEC §11.1 required tracker adapter operations.
#[async_trait]
pub trait Tracker: Send + Sync {
    /// Return issues currently in `active_states` for the configured project.
    async fn fetch_candidate_issues(&self) -> Result<Vec<Issue>, TrackerError>;

    /// Return issues whose state matches any of `state_names`. Used for the
    /// startup terminal-workspace cleanup sweep.
    async fn fetch_issues_by_states(
        &self,
        state_names: &[String],
    ) -> Result<Vec<Issue>, TrackerError>;

    /// Refresh state for a specific list of issue IDs. Used by reconciliation.
    async fn fetch_issue_states_by_ids(
        &self,
        issue_ids: &[String],
    ) -> Result<Vec<IssueState>, TrackerError>;
}

File: go/internal/tracker/tracker.go
// Package tracker defines the issue-tracker abstraction (SPEC §5.3.1).
//
// Each tracker kind (linear, github) is a separate implementation behind
// the [Tracker] interface; the orchestrator stays decoupled from any one
// vendor.
package tracker

import (
	"context"

	"github.com/noeljackson/symphony/go/internal/issue"
)

// Tracker is the read-only-from-orchestrator-perspective interface every
// tracker implementation satisfies. Mutations (issue state transitions,
// comments) are performed by the agent backend via tools, not by the
// orchestrator.
type Tracker interface {
	// FetchCandidateIssues returns issues whose tracker state intersects the
	// configured ActiveStates. Implementations apply the v3 §5.3.1 active/
	// terminal-state rules at the source where possible (for example, Linear
	// passes a state filter to its GraphQL query).
	FetchCandidateIssues(ctx context.Context) ([]issue.Issue, error)

	// FetchIssueStatesByIDs is used by reconciliation to refresh in-flight
	// issues' state. Returns the matching issues with their current state.
	FetchIssueStatesByIDs(ctx context.Context, ids []string) ([]issue.Issue, error)

	// FetchIssuesByStates returns issues whose state matches any value in
	// `states`. Used at startup for terminal-workspace cleanup (§8.6).
	FetchIssuesByStates(ctx context.Context, states []string) ([]issue.Issue, error)
}
